src/transport/builder/umessagebuilder.rs

```diff
@@ -0,0 +1,733 @@
+    ///                          with the corresponding attribute values.
+    ///
+    /// # Examples
+    ///
+    /// ```rust
+    /// # use up_rust::uprotocol::SerializationError;
+    /// use up_rust::uuid::builder::UUIDBuilder;
+    /// use up_rust::uprotocol::{UMessageType, UPriority, UUri};
+    /// use up_rust::transport::builder::UMessageBuilder;
+    ///
+    /// # fn main() -> Result<(), Box<dyn std::error::Error>> {
+    /// let method_to_invoke = UUri::try_from("my-vehicle/cabin/1/rpc.unlock_doors")?;
+    /// let reply_to_address = UUri::try_from("my-cloud/dashboard/1/rpc.response")?;
+    /// let request_id = UUIDBuilder::new().build();
+    /// let request_message = UMessageBuilder::request(&method_to_invoke, &reply_to_address, 5000)
+    ///                            .with_request_id(&request_id)
+    ///                            .build()?;
+    ///
+    /// let response_message = UMessageBuilder::response_for_request(&request_message.attributes).build()?;
+    /// assert_eq!(response_message.attributes.type_, UMessageType::UMESSAGE_TYPE_RESPONSE.into());
+    /// assert_eq!(response_message.attributes.priority, UPriority::UPRIORITY_CS1.into());
+    /// assert_eq!(response_message.attributes.source, Some(method_to_invoke).into());
+    /// assert_eq!(response_message.attributes.sink, Some(reply_to_address).into());
+    /// assert_eq!(response_message.attributes.reqid, Some(request_id).into());
+    /// # Ok(())
+    /// # }
+    /// ```
+    pub fn response_for_request(request_attributes: &'a UAttributes) -> UMessageBuilder<'a> {
+        UMessageBuilder {
+            validator: Box::new(crate::transport::validator::ResponseValidator),
+            message_type: UMessageType::UMESSAGE_TYPE_RESPONSE,
+            source: request_attributes.sink.as_ref(),
+            sink: request_attributes.source.as_ref(),
+            request_id: request_attributes.reqid.as_ref(),
+            ..Default::default()
+        }
+    }
+
+    /// Sets the message's payload.
+    ///
+    /// # Arguments
+    ///
+    /// * `payload` - The data to set as payload.
+    /// * `format` - The payload format.
+    ///
+    /// # Returns
+    ///
+    /// The builder.
+    ///
+    /// # Examples
+    ///
+    /// ```rust
+    /// # use up_rust::uprotocol::SerializationError;
+    /// use up_rust::uprotocol::{UMessageType, UPayloadFormat, UPriority, UUri};
+    /// use up_rust::transport::builder::UMessageBuilder;
+    ///
+    /// # fn main() -> Result<(), Box<dyn std::error::Error>> {
+    /// let topic = UUri::try_from("my-vehicle/cabin/1/doors.driver_side#status")?;
+    /// let message = UMessageBuilder::publish(&topic)
+    ///                 .with_payload("locked".into(), UPayloadFormat::UPAYLOAD_FORMAT_TEXT)
+    ///                 .build()?;
+    /// assert!(message.payload.is_some());
+    /// # Ok(())
+    /// # }
+    /// ```
+    pub fn with_payload(
```

Comment:
WDYT about offering another function as well, e.g. `with_protobuf_payload()` where we put a trait bound on the input parameter `payload` s.t. it must be serializable using the `protobuf` crate?

I think that could be handy to offer that service to within the function do the serialization as well as an option.

Reply:
I had thought about that as well. We already have types from the `protobuf` crate in the public API so it shouldn't be an issue.

Reply:
Okay. Darn. Yeah, let's leave it for now.

Reply:
OK. Here's an alternative proposal. What about making a `build_with_payload()` which also takes an input that is bound to be serializable via `Protobuf` returns a `Result`?

Maybe a stretch, but that'd let us allow it to fail, without it being in the middle of the chain.

WDYT?

Reply:
Not sure if I understand. Can you provide some example code?

Reply:
Sure! Something like the below. I didn't compile it, so I may have missed a thing or two. Only thing is that returning a ValidationError in that case seems a little odd.

May also be possible to factor out some of the common code (e.g. reqid, attributes stuff) between this and the `build()` we talked about turning `build_with()` into.

WDYT?

```rust
/// Creates the message based on the builder's state, including a `Profobuf` serializable object.
///
/// # Arguments
///
/// `uuid_builder` - A (shared) builder for creating UUIDs.
/// `payload` - A payload which is serializable using `Protobuf`
///
/// # Returns
///
/// A message ready to be sent using [`crate::transport::datamodel::UTransport::send`].
///
/// # Errors
///
/// Returns an error if the properties set on the builder do not represent a valid set of [`UAttributes`] or
/// there is an issue in serializing the payload using `Protobuf`
pub fn build_with_payload<T: Message>(&self, uuid_builder: &UUIDBuilder, payload: T) -> Result<UMessage, ValidationError> {
        let reqid = self.request_id.cloned().or_else(|| {
            if self.message_type == UMessageType::UMESSAGE_TYPE_REQUEST {
                Some(uuid_builder.build())
            } else {
                None
            }
        });
        let attributes = UAttributes {
            id: Some(uuid_builder.build()).into(),
            type_: self.message_type.into(),
            source: self.source.cloned().into(),
            sink: self.sink.cloned().into(),
            priority: self.priority.into(),
            ttl: self.ttl,
            token: self.token.cloned(),
            permission_level: self.permission_level,
            commstatus: self.comm_status,
            reqid: reqid.into(),
            ..Default::default()
        };
        let buf = payload
            .write_to_bytes()
            .map_err(|_e| ValidationError::new("Failed to serialize payload to bytes"))?;
        self.validator.validate(&attributes).map(|_| {
            let payload = buf
                .as_ref()
                .map(|bytes| Some(Data::Value(bytes.to_vec())))
                .map(|data| UPayload {
                    format: self.payload_format.into(),
                    data,
                    ..Default::default()
                });
            UMessage {
                attributes: Some(attributes).into(),
                payload: payload.into(),
                ..Default::default()
            }
        })
    }
}
```

Reply:
Ah, ok, now I see it. This will need to be the final step then. If we want to go that way then I would propose to change the `with_payload` function to `build_with_payload` accordingly. WDYT?

Reply:
Yup, makes sense!